### PR TITLE
perf(lsp): cleanup workspace settings scopes

### PR DESCRIPTION
**Pull request by nayeemrmn**
Before we were fetching and storing `workspace/configuration` for every opened document. But VSCode (for example) only respects differing `.vscode/settings.json` files per workspace folder. 

This removes `SpecifierSettings` and simplifies our settings storage to the following:
```rust
pub struct Settings {
  pub unscoped: WorkspaceSettings,
  pub by_workspace_folder: Option<BTreeMap<ModuleSpecifier, WorkspaceSettings>>,
}
```

Initially, the `unscoped` field is populated with whatever was passed in the LSP init options. After that and on each config change event, if the client supports `workspace/configuration` requests, we fetch settings scoped for each workspace folder plus another 'unscoped' one to use as a fallback and to replace the stale init options.

The usage of these is abstracted under a `Config::workspace_settings_for_specifier()` method.

### CODE REVIEW

File: test_util/src/lsp.rs [hunk 1]
```diff
@@ -686,21 +670,70 @@ impl LspClient {
...
-    config: Value,
+    mut config: Value,
   ) {
-    let mut builder = InitializeParamsBuilder::new();
+    let mut builder = InitializeParamsBuilder::new(config.clone());
     builder.set_root_uri(self.context.temp_dir().uri());
     do_build(&mut builder);
     let params: InitializeParams = builder.build();
+    // `config` must be updated to account for the builder changes.
+    // TODO(nayeemrmn): Maybe remove builder.
```
Comment (dsherret) [inline test_util/src/lsp.rs:718]: Why maybe remove the builder?

**Review by dsherret — APPROVED** [review]
LGTM!

File: test_util/src/lsp.rs (hunk 1 above)
Comment (nayeemrmn) [reply test_util/src/lsp.rs:718]: There's too many ways of supplying config, there's a separate arg for that and the value has to be saved to keep supplying it on configuration requests. The builder makes it complicated

Comment (nayeemrmn) [reply test_util/src/lsp.rs:718]: Oh right, yeah not the whole builder but just make the `InitializeParams::initializationOptions` part of it (which accepts workspace settings) into one value setter.

**Pull request merged**